### PR TITLE
Correct types of format argument allocation

### PR DESCRIPTION
**Pull request by romainbrenguier**
We were passing the wrong type for the allocation of an object in the string preprocessing.
To ensure this does not happen again I changed the method `allocate_dynamic_object_with_decl` so that it infers itself what type should be allocated.

### CODE REVIEW

File: regression/strings-smoke-tests/java_format2/test.desc [hunk 1]
```diff
@@ -1,8 +1,9 @@
 CORE
 test.class
---refine-strings --string-max-length 20
+--refine-strings --string-max-length 20 --verbosity 9
 ^EXIT=10$
 ^SIGNAL=0$
 ^\[.*assertion.1\].* line 6.* SUCCESS$
 ^\[.*assertion.2\].* line 7.* FAILURE$
 --
+^warning
```
Comment (owen-jones-diffblue) [inline regression/strings-smoke-tests/java_format2/test.desc]: Missing new line

File: src/java_bytecode/java_object_factory.cpp [hunk 2]
```diff
@@ -151,7 +151,7 @@ class java_object_factoryt
 /// Generates code for allocating a dynamic object. This is used in
 /// allocate_object and for allocating strings in the library preprocessing.
 /// \param target_expr: expression to which the necessary memory will be
-///   allocated
+///   allocated, its type should be pointer to `allocate_type`
 /// \param allocate_type: type of the object allocated
```
Comment (owen-jones-diffblue) [inline src/java_bytecode/java_object_factory.cpp]: allocate_type is not a parameter any more

File: src/java_bytecode/java_object_factory.cpp [hunk 3]
```diff
@@ -151,7 +151,7 @@ class java_object_factoryt
 /// Generates code for allocating a dynamic object. This is used in
 /// allocate_object and for allocating strings in the library preprocessing.
 /// \param target_expr: expression to which the necessary memory will be
-///   allocated
+///   allocated, its type should be pointer to `allocate_type`
```
Comment (owen-jones-diffblue) [inline src/java_bytecode/java_object_factory.cpp]: Now that allocate_type is not a parameter, I'm not sure this comment is correct (or rather, it is trivially correct)

Comment (romainbrenguier) [inline src/java_bytecode/java_object_factory.cpp]: It's now longer a parameter of `allocate_dynamic_object_with_decl` but is still used for `allocate_dynamic_object` where it is useful if a typecast is required.

Comment (owen-jones-diffblue) [inline src/java_bytecode/java_object_factory.cpp]: I feel this comment is backwards. It's tautological: you define `allocate_type` so that this condition is satisfied. Maybe rephrase as "we call the type that it points to `allocate_type`".

File: regression/strings-smoke-tests/java_format3/test.desc [hunk 4]
```diff
@@ -1,8 +1,9 @@
 CORE
 test.class
---refine-strings --string-max-length 20
+--refine-strings --string-max-length 20 --verbosity 9
```
Comment (owen-jones-diffblue) [inline regression/strings-smoke-tests/java_format3/test.desc]: Do we want this verbosity or is it left over from debugging?

File: regression/strings-smoke-tests/java_format/test.desc [hunk 5]
```diff
@@ -1,8 +1,9 @@
 CORE
 test.class
---refine-strings --string-max-length 20
+--refine-strings --string-max-length 20 --verbosity 9
```
Comment (owen-jones-diffblue) [inline regression/strings-smoke-tests/java_format/test.desc]: Do we want this verbosity or is it left over from debugging?

File: regression/strings-smoke-tests/java_format2/test.desc [hunk 6]
```diff
@@ -1,8 +1,9 @@
 CORE
 test.class
---refine-strings --string-max-length 20
+--refine-strings --string-max-length 20 --verbosity 9
```
Comment (owen-jones-diffblue) [inline regression/strings-smoke-tests/java_format2/test.desc]: Do we want this verbosity or is it left over from debugging?

File: regression/strings-smoke-tests/java_format3/test.desc (hunk 4 above)
Comment (romainbrenguier) [inline regression/strings-smoke-tests/java_format3/test.desc]: It's so that we can check that no warning is given. With verbosity lower, no warning is present in the output.

File: src/java_bytecode/java_object_factory.cpp (hunk 3 above)
Comment (romainbrenguier) [inline src/java_bytecode/java_object_factory.cpp]: I think  you are talking about `allocate_dynamic_object_with_decl` , but this one still has `allocate_type` as an argument

File: src/java_bytecode/java_object_factory.cpp [hunk 7]
```diff
@@ -216,32 +216,27 @@ exprt allocate_dynamic_object(
...
   const source_locationt &loc,
-  code_blockt &output_code,
-  bool cast_needed)
+  code_blockt &output_code)
 {
   std::vector<const symbolt *> symbols_created;
-
   code_blockt tmp_block;
-  exprt result=allocate_dynamic_object(
+  const typet &allocate_type=target_expr.type().subtype();
```
Comment (allredj) [inline src/java_bytecode/java_object_factory.cpp]: Is this the only functional change? It's hard to see there is a lot of refactoring in the same commit.

Comment (romainbrenguier) [inline src/java_bytecode/java_object_factory.cpp]: yes that's right

File: src/java_bytecode/java_object_factory.cpp (hunk 3 above)
Comment (owen-jones-diffblue) [inline src/java_bytecode/java_object_factory.cpp]: Indeed, I was bamboozled by github's intelligent code folding

File: regression/strings-smoke-tests/java_format3/test.desc (hunk 4 above)
Comment (peterschrammel) [inline regression/strings-smoke-tests/java_format3/test.desc]: Warnings are shown with verbosity>=2.

File: src/java_bytecode/java_object_factory.cpp (hunk 7 above)
Comment (peterschrammel) [inline src/java_bytecode/java_object_factory.cpp]: Separating into two commits would be helpful.

Comment (romainbrenguier) [inline src/java_bytecode/java_object_factory.cpp]: yes I will try to do that next time. it just seemed natural  to remove the argument made unused by this change in the same commit.
